Description: <global update>
 TODO: Meilleur prise en charge des nombres et mise à jour de la page de json2txt.3
 .
 json2txt (0.3-1) unstable; urgency=medium
 .
   * Initial release (Closes: #2024)  <2024 is the bug number of your ITP>
Author: Thomas Marsaleix <zoeurk@gmail.com>
Bug-Debian: https://bugs.debian.org/2024

---
The information above should follow the Patch Tagging Guidelines, please
checkout http://dep.debian.net/deps/dep3/ to learn about the format. Here
are templates for supplementary fields that you might want to add:

Origin: <vendor|upstream|other>, <url of original patch>
Bug: <url in upstream bugtracker>
Bug-Debian: https://bugs.debian.org/<bugnumber>
Bug-Ubuntu: https://launchpad.net/bugs/<bugnumber>
Forwarded: <no|not-needed|url proving that it has been forwarded>
Reviewed-By: <name and email of someone who approved the patch>
Last-Update: 2024-09-28

--- json2txt-0.3.orig/lib/json2txt.c
+++ json2txt-0.3/lib/json2txt.c
@@ -70,7 +70,7 @@ void *getint(struct json_parser *p){
 	int i, dot  = 0, start = 0, zero = 0;
 	char *fboolean[2] = { "false", "FALSE" },
 		*tboolean[2] = { "true", "TRUE" },
-		*pbool = NULL, *Pbool = NULL;
+		*pbool = NULL, *Pbool = NULL, last;
 	p->len = 0;
 	if(*p->buf == 't' || *p->buf == 'T'){
 		pbool = tboolean[0];
@@ -108,32 +108,33 @@ void *getint(struct json_parser *p){
 		while(read_fn(p));
 		return p;
 	}
-	#ifdef EXPLICIT_SIGN
 		if(*p->buf == '-' || *p->buf == '+'){
-			STOCK_BUF(p);
-			p->len++;
-			p->buf++;
-			p->offset++;
+			if(*p->buf == '+'){
+				warnx("Value number: start by '+', this value is not valid (offset: %lu)", p->offset);
+			#ifndef EXPLICIT_SIGN
+				errx(255, "Invalid Number");
+			#endif
+			}
+			#ifdef EXPLICIT_SIGN
+				last = *p->buf;
+				STOCK_BUF(p);
+				p->len++;
+				p->buf++;
+				p->offset++;
+			#endif
 	}
-	#else
-		if(*p->buf == '-'){
-			STOCK_BUF(p);
-			p->len++;
-			p->buf++;
-			p->offset++;
-		}
-	#endif
 	do
-		for(;*p->buf;p->buf++, p->len++, p->offset++)
+		for(;(last = *p->buf);p->buf++, p->len++, p->offset++)
 			if(*p->buf == '.'){
 				switch(dot){
 					case 0:
-						#ifdef STRICT_NUM
 						if(zero == 0 && start == 0){
-							warnx("Invalid number.");
+							warnx("Value number: start by '(+|-)?.num', valid value should be '-?0.num' (offset: %lu)", p->offset);
+							#ifdef STRICT_NUM
+							errx("Invalid number.");
 							return p;
+							#endif
 						}
-						#endif
 						dot = 1;
 						start = 1;
 						STOCK_BUF(p);
@@ -186,6 +187,7 @@ void *array(struct json_parser *p, struc
 	size_t index = 0;
 	ssize_t offset = p->offset;
 	int next = 1, end = 0;
+	char c_char;
 	p->buf++;
 	p->offset++;
 	do
@@ -254,9 +256,17 @@ void *array(struct json_parser *p, struc
 					end = 1;
 					next = 0;
 					getint(p);
+					c_char = *(p->pstock -1);
+					if(c_char == '.' || c_char == '+' || c_char == '-'){
+						errx(255, "Unexpected chararcter (bad number) before offset %lu.", p->offset);
+					}
+					c_char = *p->stock;
+					if(c_char == '+' || *(p->stock + (c_char == '+' || c_char == '-')) == '.')
+						(*j)->t_val = INT | WARN;
+					else
+						(*j)->t_val = INT;
 					(*j)->value.name.index = index;
 					(*j)->value.value = p->stock;
-					(*j)->t_val = INT;
 					index++;
 					p->stock = NULL;
 					p->stock_size = 0;
@@ -271,7 +281,7 @@ void *pair(struct json_parser *p, struct
 	struct json *pj;
 	ssize_t offset = p->offset;
 	int need_key = 1, need_value = 0,
-		next = 0, end = 0;
+		next = 0, end = 0, c_char;
 	p->offset++;
 	p->buf++;
 	do
@@ -365,8 +375,16 @@ void *pair(struct json_parser *p, struct
 					end = 1;
 					next = 0;
 					getint(p);
+					c_char = *(p->pstock -1);
+					if(c_char == '.' || c_char == '+' || c_char == '-'){
+						errx(255, "Unexpected chararcter (bad number) before offset %lu.", p->offset);
+					}
+					c_char = *p->stock;
+					if(c_char == '+' || *(p->stock + (c_char == '+' || c_char == '-')) == '.')
+						(*j)->t_val = INT | WARN;
+					else
+						(*j)->t_val = INT;
 					(*j)->value.value = p->stock;
-					(*j)->t_val = INT;
 					p->stock = NULL;
 					p->stock_size = 0;
 					break;
@@ -437,19 +455,34 @@ int duplicate_keys(struct json *j, int w
 	struct json *pj = j;
 	char **order = NULL;
 	for(pj = j; pj;pj = pj->next){
-		if(pj->type == ARRAY)
-			continue;
-		if(pj->sub)
-			if(duplicate_keys(pj->sub, warning_only) == 1){
+		if(pj->type == ARRAY){
+			if(pj->sub){
+				if(duplicate_keys(pj->sub, warning_only) == 1)
+					return 1;
+			}
+			if(warning_only >= 0 && (pj->t_val&WARN) == WARN){
+				warnx("Mal formed number: %s", pj->value.value);
+				if(warning_only == 0)
+					return 1;
+			}
+		}else{
+			if(pj->sub)
+				if(duplicate_keys(pj->sub, warning_only) == 1){
+					free(order);
+					return 1;
+				}
+			if(warning_only >= 0 && (pj->t_val&WARN) == WARN){
+				warnx("Mal formed number: %s", pj->value.value);
+				if(warning_only == 0)
+					return 1;
+			}
+			if(json_sort(pj, &order, 0, warning_only) == -1){
 				free(order);
 				return 1;
 			}
-		if(json_sort(pj, &order, 0, warning_only) == -1){
 			free(order);
-			return 1;
+			order = NULL;
 		}
-		free(order);
-		order = NULL;
 	}
 	return 0;
 }
@@ -505,7 +538,7 @@ void json_print(struct json *j, int sort
 			switch(pj->type^SET){
 				case ARRAY:
 					switch(pj->t_val){
-						case INT:
+						case INT: case INT|WARN:
 							printf("%s", pj->value.value);
 							break;
 						case STRING:
@@ -518,7 +551,7 @@ void json_print(struct json *j, int sort
 				case PAIR:
 					printf("\"%s\":", pj->value.name.key);
 					switch(pj->t_val){
-						case INT:
+						case INT: case INT|WARN:
 							printf("%s", pj->value.value);
 							break;
 						case STRING:
@@ -578,7 +611,7 @@ void json2txt(struct json *j, int sort,
 			switch(pj->type^SET){
 				case ARRAY:
 					switch(pj->t_val){
-						case INT:
+						case INT: case INT|WARN:
 							if(string)
 								printf("%s[%lu]:%s\n",
 									string, pj->value.name.index, pj->value.value);
@@ -598,7 +631,7 @@ void json2txt(struct json *j, int sort,
 					break;
 				case PAIR:
 					switch(pj->t_val){
-						case INT:
+						case INT:case INT|WARN:
 							if(string)
 								printf("%s.%s:", string, pj->value.name.key);
 							else
--- json2txt-0.3.orig/lib/json2txt.h
+++ json2txt-0.3/lib/json2txt.h
@@ -24,7 +24,8 @@ enum TYPE{
 enum VAL{
 	INT = 0,
 	STRING = 1,
-	VOID = 2
+	VOID = 2,
+	WARN = 4
 };
 enum KIND{
 	NEW,
--- json2txt-0.3.orig/man/json2txt.3
+++ json2txt-0.3/man/json2txt.3
@@ -57,8 +57,8 @@ struct json_parser
 struct json
 .EX
 \fBstruct json\fR{
-	\fBenum TYPE\fR type;					/*Type of data*/
-	\fBenum VAL\fR t_val;				/*Type of Value of the data*/
+	\fBenum TYPE\fR type;			/*Type of data*/
+	\fBenum VAL\fR t_val;			/*Type of Value of the data*/
 	struct{
 		union{
 			\fBsize_t\fR index;		/*index in array*/
@@ -87,7 +87,7 @@ Alloc/realloc buffer with lentoadd
 .BR
 .TP
 \fBvoid\fR *json_create(\fBstruct json\fR **j, \fBenum KIND\fR kind, \fBenum TYPE\fR type);
-Create a json structure by kind, and set type. *j is set to the new position and return the entering *json. 
+Create a json structure by kind, and set type. *j is set to the new position and return the entering *json.
 .BR
 .TP
 \fBvoid\fR *getint(\fBstruct json_parser\fR *p);
@@ -123,11 +123,15 @@ warn_only:
 .BR
 .TP
 \fBint\fR duplicate_keys(\fBstruct json\fR *j, \fBint\fR warning_only);
-Search for duplicate keys. Use json_sort().
+Search for duplicate keys and bad numbers. Use json_sort().
 .BR
 .TP
 \fBvoid\fR json_print(\fBstruct json\fR *j, \fBint\fR sort, \fBsize_t\fR space, \fBchar\fR c_sp, \fBsize_t\fR count, \fBint\fR warn_only);
 Print FD in json format. Use json_sort().
+.EX
+\fIspace\fR is the number of \fIc_sp\fR used for space.
+if \fIc_sp\fR == 0 no space are appliqued
+.EE
 .BR
 .TP
 \fBvoid\fR json2txt(\fBstruct json\fR *j, \fBint\fR sort, \fBchar\fR *string, \fBint\fR warn_only);
